lxd/migrate_instance.go

```diff
@@ -509,6 +509,12 @@ func (s *migrationSourceWs) Do(state *state.State, migrateOp *operations.Operati
 		}
 	}
 
+	// Attempt to acquire exclusive access (succeeds if instance is stopped).
```

Comment:
What happens if instance is running?

Will the issue this is attempting to fix still occur in that case?
If not then we should have a comment here explaining that to capture the knowledge.

Reply:
According to the test, there is no issue when the instance is running. Also, in that case, the lock is not acquired.

Reply:
What test was needing this? I'm struggling to review this because I don't know what scenario it is fixing I'm afraid.

Reply:
This is the reproducer which is part of the migration test:

```
pool1="zfs1"
lxc init images:alpine/3.17 cccp -s "${pool1}"
lxc storage volume set "${pool1}" container/cccp user.foo=snap0
echo "before" | lxc file push - cccp/blah
[ "$(lxc file pull cccp/blah -)" = "before" ]
lxc snapshot cccp
lxc storage volume set "${pool1}" container/cccp user.foo=snap1
lxc snapshot cccp
echo "after" | lxc file push - cccp/blah
[ "$(lxc file pull cccp/blah -)" = "after" ]
lxc storage volume set "${pool1}" container/cccp user.foo=postsnap1
# Local container only copy.
lxc copy cccp udssr --instance-only -s "${pool1}"
[ "$(lxc info udssr | grep -c snap)" -eq 0 ]
[ "$(lxc file pull udssr/blah -)" = "after" ]
```

The error happens only when using zfs block mode with xfs as a filesystem. In that case, the last line will result in
```
[ "before" = "after" ]
```
which is wrong.

The error is caused by the following two lines running straight after each other:
```
echo "after" | lxc file push - cccp/blah
...
lxc copy cccp udssr --instance-only -s "${pool1}"
```

The issue can be solved by sleeping for ~15 seconds between the calls. That's also the reason, users probably won't run into this issue when running the commands manually.

What really solves the issue though is the source being unmounted before doing the copy. That ensures everything is written to disk before copying.

Reply:
Right, I see now, thanks. 

And for stopped containers, the only reason its mounted is because there was a recent `lxc file push` operation.

Out of interest have you tried adding  `flush` call in the tests between `echo "after" | lxc file push - cccp/blah` and `lxc copy cccp udssr --instance-only -s "${pool1}"`?

Because I'm not sure adding an exclusive lock here is the right course of action as it will stop any ongoing file transfers and/or break any `lxc file mount` operations that are ongoing.

Reply:
> And for stopped containers, the only reason its mounted is because there was a recent lxc file push operation.

Correct.

> Out of interest have you tried adding flush call in the tests between echo "after" | lxc file push - cccp/blah and lxc copy cccp udssr --instance-only -s "${pool1}"?

I have tried adding a `sync` but that still causes the test to fail.

Reply:
And is this a ZFS specific thing or does it occur for other pool drivers too?

Reply:
If you could provide me a small test script that reproduces the issue I'd be interested at running this locally and see if we can figure out whats going on...

Reply:
Then at least we'd have something to add to the comment as to specifically why this is needed.

Reply:
> And is this a ZFS specific thing or does it occur for other pool drivers too?

I only stumble over it when I added tests for all filesystems for zfs block mode. I can try xfs on LVM and see what happens.

Reply:
Thanks. Yes I think with all these sorts of things it'd be worth trying to solve them "in the whole" rather than for specific drivers (even if that means just copying the same fix into each driver). But also if its only affecting a single driver, then solving something outside of the specific driver (especially if its introducing a limitation, like this one is) we should really understand the problem that we are trying to fix and how it fixes it.

Reply:
> If you could provide me a small test script that reproduces the issue I'd be interested at running this locally and see if we can figure out whats going on...

The above snippet should be enough. You only need a zfs pool with `zfs.block_mode=true` and `volume.block.filesystem=xfs`.

Reply:
The exclusive lock also prevents issues with a user attempting to `lxc start` or do another new file transfer partway through a migration.